### DISCUSSION TITLE
Problem to open/close partially window (NEW behavior)

### DISCUSSION THREAD

**Discussion by pilot1981** [RavenSystem/esp-homekit-devices#2177] [MEPLHAA Scripts]
**Detailed steps to reproduce it**
Not possible to open/close window partially 

**MEPLHAA Script**
It’s works fine to open/close totally!

{"c":{"n":"finestracucina","tz":"UTC+1,M3.4.0/2,M10.5.0/3","ot":"[192.168.6.41:514](http://192.168.6.41:514/)","ntp":"[192.168.6.41](http://192.168.6.41/)","io":[[[12,13,16],2],[[2,4,5],6]],"l":16,"b":[[2,5]]},"a":[{"t":45,"w":2,"o":18,"c":16,"f":70,"0":{"r":[[13],[12,1]]},"1":{"r":[[13,1],[12]]},"2":{"r":[[13],[12]]},"3":{"r":[[13],[12,0,1]]},"4":{"r":[[13,0,1],[12]]},"f0":[[5,0]],"f1":[[4,0]],"f2":[[5],[4]]}]}

**Used device**
Electrodragon relay IOT

**HAA version**
V 12.9.1

**iOS/iPadOS version**
17.2

**Logs**
NA

**Screen-shots or videos showing issue**
NA

I need to control window covering already using DC motors and proprietary electronic board, to open/close partially I need to add in each action an impulse con GPIO that reverse the way…it’s possibile with HAA? How?

it need to works in this way:
1. To open totally windows send 1 input to open command
2. To close totally window send 1 input to open command
3. To open partially windows send 1 input to open command and then send stop command at the right moment
4. To close partially windows send 1 input to close command and then send open command at the right moment

I think isn’t possible to configure it using JSON: you have to modify HAA source code?

**Comment by RavenSystem** [comment]
Use right actions for each: https://github.com/RavenSystem/esp-homekit-devices/wiki/window-covering#actions

If you need a more complex solution, you can use virtual switch services out of HomeKit.

**Comment by pilot1981** [comment]
Hi @RavenSystem,
I don't understand howto have an action to wait x seconds between first inching of GPIOX and second GPIOY...this could be related to percentage of desidered opening/closing...

**Reply by RavenSystem** [reply]
It is done automatically by HAA. Simply send inching for GPIOX in start actions and inching for GPIOY in stop actions.

**Reply by pilot1981** [reply]
You are saying changing action 5 and 6 stop during opening and during closing I can stop the window at the right percentage of opening/closing?

**Reply by pilot1981** [reply]
I tried adding action 5 and 6 with this JSON: {"c":{"n":"finestracucina","tz":"UTC+1,M3.4.0/2,M10.5.0/3","ot":"[192.168.6.41:514](http://192.168.6.41:514/)","ntp":"[192.168.6.41](http://192.168.6.41/)","io":[[[12,13,16],2],[[2,4,5],6]],"l":16,"b":[[2,5]]},"a":[{"t":45,"w":2,"o":18,"c":16,"f":70,"0":{"r":[[13],[12,1,1]]},"1":{"r":[[13,1,1],[12]]},"2":{"r":[[13],[12]]},"3":{"r":[[13],[12,0,1]]},"4":{"r":[[13,0,1],[12]]},"5":{"r":[[13,1,1],[12]]},"6":{"r":[[13],[12,1,1]]},"f0":[[5,0]],"f1":[[4,0]],"f2":[[5],[4]]}]}

but I have again problem with partial opening/closing...

**Reply by RavenSystem** [reply]
But remove action 2. It is in conflict with your actions 5 and 6, because action 2 is operating GPIO 13 and 12, setting both to LOW.

**Reply by pilot1981** [reply]
ok thanks, I found right JSON: {"c":{"n":"finestracucina","tz":"UTC+1,M3.4.0/2,M10.5.0/3","ot":"[192.168.6.41:514](http://192.168.6.41:514/)","ntp":"[192.168.6.41](http://192.168.6.41/)","io":[[[12,13,16],2],[[2,4,5],6]],"l":16,"b":[[2,5]]},"a":[{"t":45,"w":2,"o":18,"c":16,"f":70,"0":{"r":[[13],[12,1,1]]},"1":{"r":[[13,1,1],[12]]},"3":{"r":[[13],[12,0,1]]},"4":{"r":[[13,0,1],[12]]},"5":{"r":[[13,1,1],[12]]},"6":{"r":[[13],[12,1,1]]},"f0":[[5,0]],"f1":[[4,0]],"f2":[[5],[4]]}]}

now I have to add margin and/or correction value? I have wrong percentage of opening/closing anche totally opening/closing don't open/close 100%

where I have to start to work? from margin or from correction value?

**Reply by pilot1981** [reply]
Hi @RavenSystem,
I don't have close sensor, but with this last JSON {"c":{"n":"finestracucina","tz":"UTC+1,M3.4.0/2,M10.5.0/3","ot":"[192.168.6.41:514](http://192.168.6.41:514/)","ntp":"[192.168.6.41](http://192.168.6.41/)","io":[[[12,13,16],2],[[2,4,5],6]],"l":16,"b":[[2,5]]},"a":[{"t":45,"w":2,"o":18,"c":16,"f":70,"0":{"r":[[13],[12,1,1]]},"1":{"r":[[13,1,1],[12]]},"3":{"r":[[13],[12,0,1]]},"4":{"r":[[13,0,1],[12]]},"5":{"r":[[13,1,1],[12]]},"6":{"r":[[13],[12,1,1]]},"f0":[[5,0]],"f1":[[4,0]],"f2":[[5],[4]]}]}

I have new problem finding totally open/close position:
- when windows are closed totally, on homekit I have opened totally
- when windows are opened totally, on homekit I have HAA device closed totally

I only added actions 5 and 6 and removed action 2...any suggestions?
